Upgrade to Javalin 4

NicMcPhee opened the pull request:
This required a _lot_ more work than it did for Labs 2 and 3. The testing style was different here, using `ContextUtil` to create mock contexts. That all failed after upgrading to Javalin 4, mostly because of the need to explicitly provide a `JsonMapper` to convert between POJOs and JSON objects. @floogulinc suggested a way to deal with that, however, that got us over the hump.

Closes #19

NicMcPhee committed:
Merge branch 'upgrade-to-gradle-7' into upgrade-to-javalin-4

NicMcPhee committed:
Update Javalin to 4.3.0

NicMcPhee committed:
Update Server.java to work with Javalin 4

I had to pass in an instance of `JavalinJackson` so that the server would have a way to convert to/from JSON and POJOs.

I also had to update the paths with path param variables from the `:id` syntax to the new `{id}` syntax.

NicMcPhee committed:
Update `UserController` to Javalin 4

The `Context#queryParam` syntax changed, which required making a number of changes.

I also added error messages to all the `.check()` calls in the validators.

NicMcPhee committed:
Update `UserControllerSpec` to Javalin 4

I had to add several attributes to the `ContextUtil` mock context object used in the testing to get things to work in the new universe.

The big thing was that I had to inject was a `JavalinJackson` mapper that is now used to convert between POJOs and JSON objects. I also had to inject a `maxRequestSize` parameter that `ContextUtil` needs for all the POST requests.

Constructing a `ContextUtil` got complicated enough that I pulled that out into two helper functions.

Then every test that used to call `ContextUtil.init()` directly had to be changed to use the new setup.

NicMcPhee committed:
Remove the dependency on Guava

The switch to `Map.of()` means we no longer need the dependency on Guava, which should reduce the Dependabot noise a wee bit.

NicMcPhee committed:
Merge branch 'main' into upgrade-to-javalin-4

helloworld12321 commented on server/src/test/java/umm3601/user/UserControllerSpec.java:
```diff
@@ -125,6 +135,39 @@ public void setupEach() throws IOException {
 
+  /**
+   * Construct an instance of `ContextUtil`, which is essentially
+   * a mock context in Javalin. See `mockContext(String, Map) for
```
Quick typo fix

```suggestion
   * a mock context in Javalin. See `mockContext(String, Map)` for
```

helloworld12321 commented on server/src/test/java/umm3601/user/UserControllerSpec.java:
```diff
@@ -125,6 +135,39 @@ public void setupEach() throws IOException {
+        pathParams,
+        HandlerType.INVALID,
+        Map.ofEntries(
+          entry(io.javalin.plugin.json.JsonMapperKt.JSON_MAPPER_KEY, new JavalinJackson()),
```
Is it possible to import `io.javalin.plugin.json.JsonMapperKt`, so we can use the non-fully-qualified name `JsonMapperKt.JSON_MAPPER_KEY` here?

NicMcPhee replied:
Absolutely! I meant to do that, and obviously didn't clean that bit up.

NicMcPhee replied:
That should be taken care of now.

helloworld12321 commented on server/src/test/java/umm3601/user/UserControllerSpec.java:
```diff
@@ -125,6 +135,39 @@ public void setupEach() throws IOException {
+        HandlerType.INVALID,
+        Map.ofEntries(
+          entry(io.javalin.plugin.json.JsonMapperKt.JSON_MAPPER_KEY, new JavalinJackson()),
+          entry(ContextUtil.maxRequestSizeKey, maxRequestSize)));
```
What would you think about referencing this directly from JavalinConfig, rather than hardcoding it? It's a bit of extra work down here, but it's a bit DRYer and less brittle (and saves us from having that long discussion up at the top 🙂

(I agree with you that it would be nicer if `JavalinConfig` had a better API for these things 😛)

```suggestion
          // Use the default maximum request size.
          entry(ContextUtil.maxRequestSizeKey, new JavalinConfig().maxRequestSize)));
```

NicMcPhee replied:
In principle I would _totally_ prefer to refer to JavalinConfig. Unfortunately we'd have to construct an object, extract that value, and then make that available, all of which is hassle. I'd rather put that energy into just getting rid of this whole `ContextUtil` business so we don't need that constant at all.

NicMcPhee replied:
But of course I construct a `JavalinJackson` every time this gets called, which also makes no sense. I'm fixing that, and I'll move your suggestion so that it only happens once.

NicMcPhee replied:
So I pull this and my construction of `JavalinJackson` up to static fields so they should only happen once now.

helloworld12321 commented on server/src/test/java/umm3601/user/UserControllerSpec.java:
```diff
@@ -45,6 +47,16 @@
 */
 public class UserControllerSpec {
 
+  /*
+   * We need this because `ContextUtil` (essentially mock contexts) uses
+   * an internal `maxRequestSize` attribute that we have to set when we
+   * initialize a mock context. If we don't, you get an obscure
+   * `NullPointerException` when it tries to read its `maxRequestSize`
+   * attribute. The value provided here is the default that Javalin
+   * uses, at least in February 2022.
+   * (https://github.com/tipsy/javalin/blob/00152f1f78862e57bb296f34fee694a7cf47b453/javalin/src/main/java/io/javalin/core/JavalinConfig.java#L65)
+   */
+  private static final long maxRequestSize = 1_000_000L;
```
(See above)

```suggestion
```

helloworld12321 commented on server/src/test/java/umm3601/user/UserControllerSpec.java:
```diff
@@ -31,12 +32,13 @@
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 
+import io.javalin.core.validation.ValidationException;
```
(See above)

```suggestion

import io.javalin.core.JavalinConfig;
import io.javalin.core.validation.ValidationException;
```

helloworld12321 reviewed (changes requested):
A couple of suggestions—nothing major. Feel free to disagree with anything if you want! 🙂

Thanks for debugging all of this—I hear it was not a fun process.

NicMcPhee committed:
Update server/src/test/java/umm3601/user/UserControllerSpec.java

Fixes a typo – thanks to @helloworld12321!

Co-authored-by: Joe Moonan Walbran <walbr037@morris.umn.edu>

NicMcPhee committed:
DRY up `maxRequestSize` mess in `UserControllerSpec`

@helloworld12321 rightfully pushed me to get rid of the duplication in providing a separate value here for `maxRequestSize`, and showed that it really wouldn't be that big a deal to grab it from `JavalinConfig`.

This incorporates his suggestions, and moves the construction of `JavalinJackson` to a field so it only happens once as well.

Co-authored-by: Joe Moonan Walbran <walbr037@morris.umn.edu>

floogulinc commented on server/src/main/java/umm3601/user/UserController.java:
```diff
@@ -99,8 +99,14 @@ public void getUsers(Context ctx) {
 
-    String sortBy = ctx.queryParam("sortby", "name"); //Sort by sort query param, default is name
-    String sortOrder = ctx.queryParam("sortorder", "asc");
+    String sortBy = ctx.queryParam("sortby"); //Sort by sort query param, default is name
+    if (sortBy == null) {
+      sortBy = "name";
+    }
+    String sortOrder = ctx.queryParam("sortorder");
+    if (sortOrder == null) {
+      sortOrder = "asc";
+    }
```
It's too bad they removed the default value from `ctx.queryParam`. This really highlights how clunky Java can be! In Kotlin this could simply be 

```kotlin
val sortBy = ctx.queryParam("sortby") :? "name"
```

Though actually as I just found, in Java > 9 there is a way make this a bit nicer using [Objects#requireNonNullElse](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Objects.html#requireNonNullElse(T,T)):
```suggestion
    String sortBy = Objects.requireNonNullElse(ctx.queryParam("sortby"), "name"); //Sort by sort query param, default is name
    String sortOrder = Objects.requireNonNullElse(ctx.queryParam("sortorder"), "asc");
```

NicMcPhee replied:
Nice find! That code was pretty unattractive and I was actually thinking about writing a little helper function that basically does what `requireNonNullElse` presumably does.

floogulinc commented on server/src/main/java/umm3601/Server.java:
```diff
@@ -10,12 +10,12 @@
 public class Server {
 
-  static String appName = "CSCI 3601 Iteration Template";
+  static String appName = "CSci 3601 Lab 4";
```
I'm pretty sure this isn't used anywhere now so it could be removed.
```suggestion
```

floogulinc commented on server/src/main/java/umm3601/Server.java:
```diff
@@ -39,6 +39,7 @@ public static void main(String[] args) {
 
     Javalin server = Javalin.create(config -> {
       config.registerPlugin(new RouteOverviewPlugin("/api"));
+      config.jsonMapper(new JavalinJackson());
```
This can be omitted, Javalin will automatically use `new JavalinJackson()` for the JSON mapper when one isn't provided. See https://github.com/tipsy/javalin/blob/bff382795d9fb40843c6a213952fc2191c24441e/javalin/src/main/java/io/javalin/core/JavalinConfig.java#L226

```suggestion
```

NicMcPhee commented on server/src/test/java/umm3601/user/UserControllerSpec.java:
```diff
@@ -55,7 +59,8 @@
   static MongoDatabase db;
 
-  static ObjectMapper jsonMapper = new ObjectMapper();
+  private static ObjectMapper objectMapper = new ObjectMapper();
```
Thanks for raising this. I had a mental note to at some point figure out if I could just have one of these, and then forgot to look into it.

NicMcPhee replied:
This should be taken care of in [83200f4](https://github.com/UMM-CSci-3601/3601-lab4-mongo/pull/22/commits/83200f4a459c441248e16528b83d5132a92b3f82).

NicMcPhee commented on server/src/test/java/umm3601/user/UserControllerSpec.java:
```diff
@@ -125,6 +128,39 @@ public void setupEach() throws IOException {
+        pathParams,
+        HandlerType.INVALID,
+        Map.ofEntries(
+          entry(JsonMapperKt.JSON_MAPPER_KEY, javalinJackson),
```
After changing the import as suggested by @floogulinc, we can just refer directly to `JSON_MAPPER_KEY`.

```suggestion
          entry(JSON_MAPPER_KEY, javalinJackson),
```

NicMcPhee committed:
Apply clean-up suggestions from @floogulinc

This incorporates a collection of nice "clean up" changes from @floogulinc.

Co-authored-by: Paul Friederichsen <floogulinc@gmail.com>

NicMcPhee committed:
Add missing import to `UserController.java`

The dangers of "in place" commits in GitHub; I missed the need to import `Objects`, but it was clear as soon as anything tried to compile the code. :)

NicMcPhee committed:
Remove redundant `ObjectMapper` from `UserControllerSpec`

@floogulinc pointed out that we really didn't need both an `ObjectMapper` and a `JavalinJackson` object since they both do essentially the same thing. This eliminates the `objectMapper` and replaces calls to it with calls to the `javalinJackson` object.

NicMcPhee committed:
Clean up the imports

The new `static` import of `JSON_MAPPER_KEY` was in the wrong place, and the import of `ObjectMapper` was no longer necessary.

floogulinc commented on server/src/main/java/umm3601/Server.java:
```diff
@@ -10,12 +10,11 @@
 import io.javalin.Javalin;
 import io.javalin.core.util.RouteOverviewPlugin;
-
+import io.javalin.plugin.json.JavalinJackson;
```
This import is now unused I think.
```suggestion
```

floogulinc reviewed (commented):
Just one small unused import but otherwise LGTM.

NicMcPhee committed:
Remove an unused import from `Server.java`

One last unnecessary import had snuck through.

Co-authored-by: Paul Friederichsen <floogulinc@gmail.com>

helloworld12321 reviewed (approved):
Looks excellent, thanks for all your work here!
